compiler/src/dotty/tools/dotc/ast/Positioned.scala

```diff
@@ -51,8 +51,12 @@ abstract class Positioned(implicit @constructorOnly src: SourceFile) extends Pro
     else {
       val newpd: this.type =
         if (mySpan.isSynthetic) {
-          if (!mySpan.exists && span.exists)
+          if (!mySpan.exists && span.exists) {
             envelope(source, span.startPos) // fill in children spans
+            () // Note: the `()` is there to prevent some inefficient code from being generated.
+               // Without it we get an allocation of a span here since the result type of the `if`
```

Comment:
I don't think this changes whether an allocation happens or not but it does make the code more efficient, I've opened an issue to track this that you could refer to in this comment: https://github.com/lampepfl/dotty/issues/5750